<a href="https://colab.research.google.com/github/AnupKaphle58/covid-research/blob/main/DataCleaning_Integration.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [81]:
import pandas as pd

# Download the file and load into a DataFrame
url = "https://catalog.ourworldindata.org/garden/covid/latest/compact/compact.csv"
covid_df = pd.read_csv(url)

# Save to local path if needed
covid_df.to_csv('/content/compact_covid.csv', index=False)

# Optional: To download it to your computer
from google.colab import files
files.download('/content/compact_covid.csv')


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [82]:
import pandas as pd
import numpy as np

df=pd.read_csv('/content/compact_covid.csv')
df.head()

country        date  total_cases  new_cases  new_cases_smoothed  \
0  Afghanistan  2020-01-01          NaN        NaN                 NaN   
1  Afghanistan  2020-01-02          NaN        NaN                 NaN   
2  Afghanistan  2020-01-03          NaN        NaN                 NaN   
3  Afghanistan  2020-01-04          0.0        0.0                 NaN   
4  Afghanistan  2020-01-05          0.0        0.0                 NaN   

   total_cases_per_million  new_cases_per_million  \
0                      NaN                    NaN   
1                      NaN                    NaN   
2                      NaN                    NaN   
3                      0.0                    0.0   
4                      0.0                    0.0   

   new_cases_smoothed_per_million  total_deaths  new_deaths  ...  population  \
0                             NaN           NaN         NaN  ...  40578801.0   
1                             NaN           NaN         NaN  ...  40578801.0   
2                             NaN           NaN         NaN  ...  40578801.0   
3                             NaN           0.0         0.0  ...  40578801.0   
4                             NaN           0.0         0.0  ...  40578801.0   

   population_density  median_age  life_expectancy  gdp_per_capita  \
0           62.215477   16.752001              NaN     1516.273315   
1           62.215477   16.752001              NaN     1516.273315   
2           62.215477   16.752001              NaN     1516.273315   
3           62.215477   16.752001              NaN     1516.273315   
4           62.215477   16.752001              NaN     1516.273315   

   extreme_poverty  diabetes_prevalence  handwashing_facilities  \
0              NaN                 10.9               48.214695   
1              NaN                 10.9               48.214695   
2              NaN                 10.9               48.214695   
3              NaN                 10.9               48.214695   
4              NaN                 10.9               48.214695   

   hospital_beds_per_thousand  human_development_index  
0                        0.39                      NaN  
1                        0.39                      NaN  
2                        0.39                      NaN  
3                        0.39                      NaN  
4                        0.39                      NaN  

[5 rows x 61 columns]

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 508524 entries, 0 to 508523
Data columns (total 61 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   country                                     508524 non-null  object 
 1   date                                        508524 non-null  object 
 2   total_cases                                 493767 non-null  float64
 3   new_cases                                   490849 non-null  float64
 4   new_cases_smoothed                          489644 non-null  float64
 5   total_cases_per_million                     493767 non-null  float64
 6   new_cases_per_million                       490849 non-null  float64
 7   new_cases_smoothed_per_million              489644 non-null  float64
 8   total_deaths                                493767 non-null  float64
 9   new_deaths                                  491630 non-null  float64
 

In [67]:
# Columns to drop
cols_to_drop = ['icu_patients_per_million', 'hosp_patients_per_million', 'excess_mortality']

# Drop columns if they exist (to avoid errors)
covid_df = covid_df.drop(columns=[col for col in cols_to_drop if col in covid_df.columns])

# Preview the result
print(covid_df.columns)
covid_df.head()

Index(['total_cases', 'new_cases', 'total_cases_per_million',
       'new_cases_per_million', 'total_deaths', 'new_deaths',
       'total_deaths_per_million', 'new_deaths_per_million',
       'diabetes_prevalence', 'cardiovasc_death_rate', 'life_expectancy',
       'population_density', 'median_age', 'aged_65_older', 'gdp_per_capita',
       'hospital_beds_per_thousand', 'stringency_index', 'iso_code',
       'continent', 'location', 'date', 'population'],
      dtype='object')


total_cases  new_cases  total_cases_per_million  new_cases_per_million  \
0          0.0        0.0                      0.0                    0.0   
1          0.0        0.0                      0.0                    0.0   
2          0.0        0.0                      0.0                    0.0   
3          0.0        0.0                      0.0                    0.0   
4          0.0        0.0                      0.0                    0.0   

   total_deaths  new_deaths  total_deaths_per_million  new_deaths_per_million  \
0           0.0         0.0                       0.0                     0.0   
1           0.0         0.0                       0.0                     0.0   
2           0.0         0.0                       0.0                     0.0   
3           0.0         0.0                       0.0                     0.0   
4           0.0         0.0                       0.0                     0.0   

   diabetes_prevalence  cardiovasc_death_rate  ...  median_age  aged_65_older  \
0                 9.59                 597.03  ...        18.6           2.58   
1                 9.59                 597.03  ...        18.6           2.58   
2                 9.59                 597.03  ...        18.6           2.58   
3                 9.59                 597.03  ...        18.6           2.58   
4                 9.59                 597.03  ...        18.6           2.58   

   gdp_per_capita  hospital_beds_per_thousand  stringency_index  iso_code  \
0         1803.99                         0.5               0.0       AFG   
1         1803.99                         0.5               0.0       AFG   
2         1803.99                         0.5               0.0       AFG   
3         1803.99                         0.5               0.0       AFG   
4         1803.99                         0.5               0.0       AFG   

   continent     location      date  population  
0       Asia  Afghanistan  1/5/2020  41128772.0  
1       Asia  Afghanistan  1/6/2020  41128772.0  
2       Asia  Afghanistan  1/7/2020  41128772.0  
3       Asia  Afghanistan  1/8/2020  41128772.0  
4       Asia  Afghanistan  1/9/2020  41128772.0  

[5 rows x 22 columns]

In [89]:
necessary_columns = [
    # COVID-19 Outcome Variables
    'total_cases', 'new_cases',  'total_cases_per_million','new_cases_smoothed',
    'new_cases_per_million', 'new_cases_smoothed_per_million','total_deaths',
    'new_deaths', 'new_deaths_smoothed', 'total_deaths_per_million',
    'new_deaths_per_million','new_deaths_smoothed_per_million',


    # Control Variables
    'population_density', 'median_age', 'diabetes_prevalence', 'gdp_per_capita',
     'stringency_index','hospital_beds_per_thousand','handwashing_facilities',

    # Additional Context Variables
    'code', 'continent', 'country', 'date', 'population'
]

# Step 4: Select the necessary columns
df_selected = df[necessary_columns]

# Step 5: Display the first 5 rows of the selected columns
print(df_selected.head())

# Step 6: Save the selected data to a new CSV file (optional)
df_selected.to_csv('selected_covid_data.csv', index=False)

   total_cases  new_cases  total_cases_per_million  new_cases_smoothed  \
0          NaN        NaN                      NaN                 NaN   
1          NaN        NaN                      NaN                 NaN   
2          NaN        NaN                      NaN                 NaN   
3          0.0        0.0                      0.0                 NaN   
4          0.0        0.0                      0.0                 NaN   

   new_cases_per_million  new_cases_smoothed_per_million  total_deaths  \
0                    NaN                             NaN           NaN   
1                    NaN                             NaN           NaN   
2                    NaN                             NaN           NaN   
3                    0.0                             NaN           0.0   
4                    0.0                             NaN           0.0   

   new_deaths  new_deaths_smoothed  total_deaths_per_million  ...  \
0         NaN                  NaN       

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 508524 entries, 0 to 508523
Data columns (total 24 columns):
 #   Column                           Non-Null Count   Dtype  
---  ------                           --------------   -----  
 0   total_cases                      493767 non-null  float64
 1   new_cases                        490849 non-null  float64
 2   total_cases_per_million          493767 non-null  float64
 3   new_cases_smoothed               489644 non-null  float64
 4   new_cases_per_million            490849 non-null  float64
 5   new_cases_smoothed_per_million   489644 non-null  float64
 6   total_deaths                     493767 non-null  float64
 7   new_deaths                       491630 non-null  float64
 8   new_deaths_smoothed              490417 non-null  float64
 9   total_deaths_per_million         493767 non-null  float64
 10  new_deaths_per_million           491630 non-null  float64
 11  new_deaths_smoothed_per_million  490417 non-null  float64
 12  po

In [92]:
# Drop columns with >50% missing data
df = df.dropna(thresh=len(df)*0.5, axis=1)

In [93]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 508524 entries, 0 to 508523
Data columns (total 22 columns):
 #   Column                           Non-Null Count   Dtype  
---  ------                           --------------   -----  
 0   total_cases                      493767 non-null  float64
 1   new_cases                        490849 non-null  float64
 2   total_cases_per_million          493767 non-null  float64
 3   new_cases_smoothed               489644 non-null  float64
 4   new_cases_per_million            490849 non-null  float64
 5   new_cases_smoothed_per_million   489644 non-null  float64
 6   total_deaths                     493767 non-null  float64
 7   new_deaths                       491630 non-null  float64
 8   new_deaths_smoothed              490417 non-null  float64
 9   total_deaths_per_million         493767 non-null  float64
 10  new_deaths_per_million           491630 non-null  float64
 11  new_deaths_smoothed_per_million  490417 non-null  float64
 12  po

this study showed handling missing data in covid dataset was better with k nearest neighbors rather than Multiple Imputation by Chained Equations (MICE)

https://link.springer.com/chapter/10.1007/978-981-19-2828-4_62?utm_source=chatgpt.com



K-Nearest Neighbors (KNN) imputation can handle both categorical and numerical data, but it requires some preprocessing. Since KNN relies on distances, categorical features must be encoded properly.

continent	✅ One-Hot Encoding	Small, unordered categories (prevents false relationships).
location	✅ Label Encoding	Unique country names (avoids high cardinality).
date	✅ Convert to DateTime & Extract Features

In [94]:
from sklearn.preprocessing import LabelEncoder
import pandas as pd

df.drop(columns=["code"], inplace=True)

df["date"] = pd.to_datetime(df["date"])
df["day_of_year"] = df["date"].dt.dayofyear
df["month"] = df["date"].dt.month
df["weekday"] = df["date"].dt.weekday

# Apply Label Encoding to location (since iso_code is dropped)
le_location = LabelEncoder()
df["country"] = le_location.fit_transform(df["country"])

# Apply One-Hot Encoding to continent
df = pd.get_dummies(df, columns=["continent"])

In [95]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 508524 entries, 0 to 508523
Data columns (total 29 columns):
 #   Column                           Non-Null Count   Dtype         
---  ------                           --------------   -----         
 0   total_cases                      493767 non-null  float64       
 1   new_cases                        490849 non-null  float64       
 2   total_cases_per_million          493767 non-null  float64       
 3   new_cases_smoothed               489644 non-null  float64       
 4   new_cases_per_million            490849 non-null  float64       
 5   new_cases_smoothed_per_million   489644 non-null  float64       
 6   total_deaths                     493767 non-null  float64       
 7   new_deaths                       491630 non-null  float64       
 8   new_deaths_smoothed              490417 non-null  float64       
 9   total_deaths_per_million         493767 non-null  float64       
 10  new_deaths_per_million           491630 non-

**ALCOHOL DATASET**

In [25]:
import pandas as pd

df = pd.read_excel('/content/alcoholdata.xlsx',skiprows=2)


print("Columns in dataset:\n", df.columns.tolist())

Columns in dataset:
 ['IndicatorCode', 'Indicator', 'ValueType', 'ParentLocationCode', 'ParentLocation', 'Location type', 'SpatialDimValueCode', 'Location', 'Period type', 'Period', 'IsLatestYear', 'Dim1 type', 'Dim1', 'Dim1ValueCode', 'Dim2 type', 'Dim2', 'Dim2ValueCode', 'Dim3 type', 'Dim3', 'Dim3ValueCode', 'DataSourceDimValueCode', 'DataSource', 'FactValueNumericPrefix', 'FactValueNumeric', 'FactValueUoM', 'FactValueNumericLowPrefix', 'FactValueNumericLow', 'FactValueNumericHighPrefix', 'FactValueNumericHigh', 'Value', 'FactValueTranslationID', 'FactComments', 'Language', 'DateModified']


In [26]:
df_selected = df[[
    'SpatialDimValueCode',  # Country Code
    'Location',             # Country Name
    'Period',               # Year
    'Dim1',                 # Alcohol Type (e.g. Beer, Wine, All types)
    'Value',                # Alcohol consumption in litres
]]

In [27]:
df_selected.columns = ['CountryCode', 'Country', 'Year', 'AlcoholType', 'AlcoholConsumption']

In [28]:
df_selected.head(10)

CountryCode Country  Year                AlcoholType  AlcoholConsumption
0         AGO  Angola  2020                  All types                 3.3
1         AGO  Angola  2020                       Beer                 1.9
2         AGO  Angola  2020                       Wine                 0.2
3         AGO  Angola  2020                    Spirits                 1.1
4         AGO  Angola  2020  Other alcoholic beverages                 0.1
5         AGO  Angola  2019                  All types                 5.3
6         AGO  Angola  2019                       Beer                 2.9
7         AGO  Angola  2019                       Wine                 0.4
8         AGO  Angola  2019                    Spirits                 1.8
9         AGO  Angola  2019  Other alcoholic beverages                 0.1

we are analyzing alcohol consumption in general (e.g., for correlation with COVID-19 deaths), then using only:

➤ AlcoholType == "All types"
is the best and most reliable choice because:



In [29]:
# Filter to include only 'All types'
df_all_types = df_selected[df_selected['AlcoholType'] == 'All types']

# Drop 'AlcoholType' column since it's now redundant
df_clean = df_all_types.drop(columns='AlcoholType')

# Rename for clarity
df_clean = df_clean.rename(columns={
    'AlcoholConsumption': 'TotalAlcoholLiters'
})


df_clean.head(21)

CountryCode Country  Year  TotalAlcoholLiters
0           AGO  Angola  2020                 3.3
5           AGO  Angola  2019                 5.3
10          AGO  Angola  2018                 5.3
15          AGO  Angola  2017                 5.8
20          AGO  Angola  2016                 6.1
25          AGO  Angola  2015                 6.9
30          AGO  Angola  2014                 7.4
35          AGO  Angola  2013                 7.9
40          AGO  Angola  2012                 7.7
45          AGO  Angola  2011                 7.7
50          AGO  Angola  2010                 7.4
55          AGO  Angola  2009                 7.2
60          AGO  Angola  2008                 6.8
65          AGO  Angola  2007                 6.7
70          AGO  Angola  2006                 6.3
75          AGO  Angola  2005                 5.7
80          AGO  Angola  2004                 5.1
85          AGO  Angola  2003                 4.6
90          AGO  Angola  2002                 4.5
95          AGO  Angola  2001                 4.2
100         AGO  Angola  2000                 4.2

calculating the average alcohol consumption from 2000 to 2019 per country will give you a solid metric for how “alcoholic” a country generally is.

In [32]:
# Filter years between 2000 and 2019 (inclusive)
df_filtered = df_clean[(df_clean['Year'] >= 2000) & (df_clean['Year'] <= 2020)]

# Group by Country and calculate mean alcohol consumption
df_avg = df_filtered.groupby(['CountryCode', 'Country'])['TotalAlcoholLiters'].mean().reset_index()

# Rename the column for clarity
df_avg = df_avg.rename(columns={'TotalAlcoholLiters': 'AvgAlcohol2000_2020'})

# Round to 2 decimal places (optional)
df_avg['AvgAlcohol2000_2020'] = df_avg['AvgAlcohol2000_2020'].round(5)

# Preview
df_avg.head(10)


CountryCode               Country  AvgAlcohol2000_2020
0         AFG           Afghanistan              0.00000
1         AGO                Angola              6.00476
2         ALB               Albania              4.39524
3         AND               Andorra             11.35714
4         ARE  United Arab Emirates              1.79524
5         ARG             Argentina              7.04286
6         ARM               Armenia              3.76667
7         ATG   Antigua and Barbuda              8.42381
8         AUS             Australia             10.00952
9         AUT               Austria             12.02381

In [34]:
output_path = '/content/avg_alcohol_2000_2020.csv'
df_avg.to_csv(output_path, index=False)

**FOR SMOKING DATASET**

In [38]:
import pandas as pd

smoking_df = pd.read_csv('/content/share-of-adults-who-smoke.csv')

smoking_df.head(40)

Entity Code  Year  Prevalence of current tobacco use (% of adults)
0   Afghanistan  AFG  2000                                             36.9
1   Afghanistan  AFG  2005                                             32.7
2   Afghanistan  AFG  2010                                             29.1
3   Afghanistan  AFG  2015                                             26.0
4   Afghanistan  AFG  2018                                             23.8
5   Afghanistan  AFG  2019                                             23.8
6   Afghanistan  AFG  2020                                             23.3
7       Albania  ALB  2000                                             35.0
8       Albania  ALB  2005                                             31.3
9       Albania  ALB  2010                                             28.0
10      Albania  ALB  2015                                             25.1
11      Albania  ALB  2018                                             23.0
12      Albania  ALB  2019                                             23.0
13      Albania  ALB  2020                                             22.4
14      Algeria  DZA  2000                                             22.2
15      Algeria  DZA  2005                                             21.8
16      Algeria  DZA  2010                                             21.5
17      Algeria  DZA  2015                                             21.1
18      Algeria  DZA  2018                                             21.0
19      Algeria  DZA  2019                                             21.0
20      Algeria  DZA  2020                                             21.0
21      Andorra  AND  2000                                             35.9
22      Andorra  AND  2005                                             34.8
23      Andorra  AND  2010                                             34.0
24      Andorra  AND  2015                                             33.0
25      Andorra  AND  2018                                             32.0
26      Andorra  AND  2019                                             32.0
27      Andorra  AND  2020                                             31.8
28    Argentina  ARG  2000                                             34.0
29    Argentina  ARG  2005                                             31.5
30    Argentina  ARG  2010                                             28.9
31    Argentina  ARG  2015                                             26.6
32    Argentina  ARG  2018                                             24.9
33    Argentina  ARG  2019                                             24.9
34    Argentina  ARG  2020                                             24.5
35      Armenia  ARM  2000                                             32.0
36      Armenia  ARM  2005                                             30.3
37      Armenia  ARM  2010                                             28.6
38      Armenia  ARM  2015                                             26.9
39      Armenia  ARM  2018                                             25.8

In [40]:
# Filter years between 2000 and 2020 (inclusive)
smoking_filtered = smoking_df[(smoking_df['Year'] >= 2000) & (smoking_df['Year'] <= 2020)]

# Group by country code and name, calculate average prevalence
smoking_avg = smoking_filtered.groupby(['Code', 'Entity'])['Prevalence of current tobacco use (% of adults)'].mean().reset_index()

# Rename columns for consistency
smoking_avg = smoking_avg.rename(columns={
    'Code': 'CountryCode',
    'Entity': 'Country',
    'Prevalence of current tobacco use (% of adults)': 'AverageSmokersAdults'
})

# Round to 2 decimals (optional)
smoking_avg['AverageSmokersAdults'] = smoking_avg['AverageSmokersAdults'].round(2)

# Preview
smoking_avg.head(10)

CountryCode      Country  AverageSmokersAdults
0         AFG  Afghanistan                 27.94
1         ALB      Albania                 26.83
2         AND      Andorra                 33.36
3         ARG    Argentina                 27.90
4         ARM      Armenia                 27.84
5         AUS    Australia                 17.29
6         AUT      Austria                 35.00
7         AZE   Azerbaijan                 25.73
8         BDI      Burundi                 15.79
9         BEL      Belgium                 26.97

Burundi (BDI) has an AverageSmokers value of 15.79, which means:

On average, 15.79% of adults in Burundi were current tobacco users (smokers) from 2000 to 2020.

In [41]:
output_path = '/content/avg_smokersadults_2000_2020.csv'
df_avg.to_csv(output_path, index=False)

**FOR OBESE DATASET**

In [45]:
# Load the obesity dataset
obese_df = pd.read_csv('/content/share-of-adults-defined-as-obese.csv')

# Preview first 5 rows
obese_df.head(65)

Entity Code  Year  \
0    Afghanistan  AFG  1975   
1    Afghanistan  AFG  1976   
2    Afghanistan  AFG  1977   
3    Afghanistan  AFG  1978   
4    Afghanistan  AFG  1979   
..           ...  ...   ...   
60  Africa (WHO)  NaN  1993   
61  Africa (WHO)  NaN  1994   
62  Africa (WHO)  NaN  1995   
63  Africa (WHO)  NaN  1996   
64  Africa (WHO)  NaN  1997   

    Prevalence of obesity among adults, BMI >= 30 (crude estimate) (%) - Sex: both sexes - Age group: 18+  years  
0                                                 0.4                                                             
1                                                 0.4                                                             
2                                                 0.5                                                             
3                                                 0.5                                                             
4                                                 0.5                                                             
..                                                ...                                                             
60                                                3.8                                                             
61                                                3.9                                                             
62                                                4.1                                                             
63                                                4.3                                                             
64                                                4.5                                                             

[65 rows x 4 columns]

In [46]:
unique_years = obese_df['Year'].unique()
unique_years.sort()

print("Available years in obesity dataset:")
print(unique_years)

Available years in obesity dataset:
[1975 1976 1977 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988
 1989 1990 1991 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002
 2003 2004 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016]


In [54]:
# Step 1: Rename using exact long column name
obese_df = obese_df.rename(columns={
    'Entity': 'Country',
    'Code': 'CountryCode',
    'Prevalence of obesity among adults, BMI >= 30 (crude estimate) (%) - Sex: both sexes - Age group: 18+  years': 'ObesityRate'
})

In [55]:
obese_df.head()

Country CountryCode  Year  ObesityRate
0  Afghanistan         AFG  1975          0.4
1  Afghanistan         AFG  1976          0.4
2  Afghanistan         AFG  1977          0.5
3  Afghanistan         AFG  1978          0.5
4  Afghanistan         AFG  1979          0.5

In [56]:
obese_filtered = obese_df[(obese_df['Year'] >= 2000) & (obese_df['Year'] <= 2016)]

# Step 3: Group and average
obese_avg = obese_filtered.groupby(['CountryCode', 'Country'])['ObesityRate'].mean().reset_index()

# Step 4: Round and preview
obese_avg['AverageObesity'] = obese_avg['ObesityRate'].round(2)
obese_avg = obese_avg.drop(columns=['ObesityRate'])

obese_avg.head()

CountryCode               Country  AverageObesity
0         AFG           Afghanistan            3.04
1         AGO                Angola            4.69
2         ALB               Albania           17.48
3         AND               Andorra           25.44
4         ARE  United Arab Emirates           23.23

Prevalence of obesity:-How common obesity is in a population (as a percentage).

Adults, BMI ≥ 30	Adults whose Body Mass Index (BMI) is 30 or above, which is the medical definition of obesity.

Crude estimate (%)	A basic, unadjusted average — not accounting for factors like age structure or urban vs rural differences. Value is given in percentage (%).

Sex: both sexes	The data includes both men and women, not separated by gender.

Age group: 18+ years	Only adults aged 18 and above are included in the calculation.

 Example:
If AverageObesity = 27.5 for a country:

On average, 27.5% of adults aged 18+ (both men and women) in that country had a BMI ≥ 30 between 2000–2016.

In [57]:
output_path = '/content/avg_obesity_2000_2016.csv'
df_avg.to_csv(output_path, index=False)

COMBINE THE OBESITY DATA,ALCOHOL DATA AND SMOKERS DATA AND MAKE IT INTO A SINGLE DATASET SUCH THAT THE DATASET HAS 3 COLOUMN BUT THEY ALL HAVE THE  SAME COUNTRIES ALL 3 MUST HAVE THE COUNTRIES OTHERWISE DONOT INCLUDE IT
Only includes countries that are present in all three datasets.

Has 3 columns:

AverageObesity

AvgAlcohol2000_2019

AverageSmokers

Along with CountryCode and Country as identifiers

In [59]:
for df in [obese_avg, df_avg, smoking_avg]:
    df['Country'] = df['Country'].str.upper()

In [60]:
for df in [obese_avg, df_avg, smoking_avg]:
    df['CountryCode'] = df['CountryCode'].str.upper()

df['CountryCode'].head()

0    AFG
1    ALB
2    AND
3    ARG
4    ARM
Name: CountryCode, dtype: object

In [61]:
# Step 1: Merge obesity and alcohol on CountryCode & Country
merged_df = pd.merge(obese_avg, df_avg, on=['CountryCode', 'Country'], how='inner')

# Step 2: Merge result with smoking data
final_df = pd.merge(merged_df, smoking_avg, on=['CountryCode', 'Country'], how='inner')

# Preview
final_df.head()

CountryCode      Country  AverageObesity  AvgAlcohol2000_2020  \
0         AFG  AFGHANISTAN            3.04              0.00000   
1         ALB      ALBANIA           17.48              4.39524   
2         AND      ANDORRA           25.44             11.35714   
3         ARG    ARGENTINA           24.34              7.04286   
4         ARM      ARMENIA           17.04              3.76667   

   AverageSmokersAdults  
0                 27.94  
1                 26.83  
2                 33.36  
3                 27.90  
4                 27.84

In [63]:
num_countries = final_df['Country'].nunique()
print(f"Number of unique countries in the merged dataset: {num_countries}")

Number of unique countries in the merged dataset: 144


In [64]:
final_df.to_csv('/content/combined_health_data.csv', index=False)

COMBINING BOTH COVID DATASET WITH HEALTH FACTORS SO THAT each input sample (daily record) includes both:

Time-varying features (cases, deaths, etc.)

Time-invariant health factors (obesity, smoking, alcohol)

In [76]:
num_unique_countries = health_df['Country'].nunique()
print(f"Number of unique countries in health_df: {num_unique_countries}")

Number of unique countries in health_df: 144


In [78]:
import pandas as pd

# Load the COVID dataset
covid_df = pd.read_csv('/content/selected_covid_data.csv')

# Get unique country names from the 'location' column
unique_locations = covid_df['location'].unique()

# Print the number of unique countries and the list itself
print(f"Number of unique locations: {len(unique_locations)}")
print("Unique locations:")
print(unique_locations)


Number of unique locations: 56
Unique locations:
['Afghanistan' 'Africa' 'Albania' 'Algeria' 'American Samoa' 'Andorra'
 'Angola' 'Anguilla' 'Antigua and Barbuda' 'Argentina' 'Armenia' 'Aruba'
 'Asia' 'Australia' 'Austria' 'Azerbaijan' 'Bahamas' 'Bahrain'
 'Bangladesh' 'Barbados' 'Belarus' 'Belgium' 'Belize' 'Benin' 'Bermuda'
 'Bhutan' 'Bolivia' 'Bonaire Sint Eustatius and Saba'
 'Bosnia and Herzegovina' 'Botswana' 'Brazil' 'British Virgin Islands'
 'Brunei' 'Bulgaria' 'Burkina Faso' 'Burundi' 'Cambodia' 'Cameroon'
 'Canada' 'Cape Verde' 'Cayman Islands' 'Central African Republic' 'Chad'
 'Chile' 'China' 'Colombia' 'Comoros' 'Congo' 'Cook Islands' 'Costa Rica'
 "Cote d'Ivoire" 'Croatia' 'Cuba' 'Curacao' 'Cyprus' 'Czechia']


In [79]:
import pandas as pd

# Load datasets
covid_df = pd.read_csv('/content/selected_covid_data.csv')
health_df = pd.read_csv('/content/combined_health_data.csv')

# Standardize capitalization
covid_df['iso_code'] = covid_df['iso_code'].str.upper()
covid_df['location'] = covid_df['location'].str.upper()
health_df['CountryCode'] = health_df['CountryCode'].str.upper()
health_df['Country'] = health_df['Country'].str.upper()

# Step 1: Get unique country codes from COVID data
covid_country_codes = covid_df['iso_code'].unique()

# Step 2: Filter health data to include only those 56 countries
health_df_filtered = health_df[health_df['CountryCode'].isin(covid_country_codes)]

# Step 3: Merge filtered health data with COVID data
merged_df = pd.merge(covid_df, health_df_filtered, left_on='iso_code', right_on='CountryCode', how='inner')

# Step 4: Preview and Save
print(f"Number of rows in merged dataset: {len(merged_df)}")
print(f"Number of unique countries in merged dataset: {merged_df['location'].nunique()}")
print(merged_df.head())

Number of rows in merged dataset: 63060
Number of unique countries in merged dataset: 38
   total_cases  new_cases  total_cases_per_million  new_cases_per_million  \
0          0.0        0.0                      0.0                    0.0   
1          0.0        0.0                      0.0                    0.0   
2          0.0        0.0                      0.0                    0.0   
3          0.0        0.0                      0.0                    0.0   
4          0.0        0.0                      0.0                    0.0   

   total_deaths  new_deaths  total_deaths_per_million  new_deaths_per_million  \
0           0.0         0.0                       0.0                     0.0   
1           0.0         0.0                       0.0                     0.0   
2           0.0         0.0                       0.0                     0.0   
3           0.0         0.0                       0.0                     0.0   
4           0.0         0.0                

In [80]:
print("Countries included in the merged dataset:")
print(sorted(merged_df['location'].unique()))

Countries included in the merged dataset:
['AFGHANISTAN', 'ALBANIA', 'ALGERIA', 'ANDORRA', 'ARGENTINA', 'ARMENIA', 'AUSTRALIA', 'AUSTRIA', 'AZERBAIJAN', 'BAHAMAS', 'BAHRAIN', 'BANGLADESH', 'BARBADOS', 'BELARUS', 'BELGIUM', 'BELIZE', 'BENIN', 'BOSNIA AND HERZEGOVINA', 'BOTSWANA', 'BRAZIL', 'BULGARIA', 'BURKINA FASO', 'BURUNDI', 'CAMBODIA', 'CAMEROON', 'CANADA', 'CHAD', 'CHILE', 'CHINA', 'COLOMBIA', 'COMOROS', 'CONGO', 'COSTA RICA', "COTE D'IVOIRE", 'CROATIA', 'CUBA', 'CYPRUS', 'CZECHIA']


**COVID DATASET SELECTED COLOUMNS PROCESSING**

In [107]:
import pandas as pd
df=pd.read_csv('/content/selected_covid_data.csv')
df.head()

total_cases  new_cases  total_cases_per_million  new_cases_smoothed  \
0          NaN        NaN                      NaN                 NaN   
1          NaN        NaN                      NaN                 NaN   
2          NaN        NaN                      NaN                 NaN   
3          0.0        0.0                      0.0                 NaN   
4          0.0        0.0                      0.0                 NaN   

   new_cases_per_million  new_cases_smoothed_per_million  total_deaths  \
0                    NaN                             NaN           NaN   
1                    NaN                             NaN           NaN   
2                    NaN                             NaN           NaN   
3                    0.0                             NaN           0.0   
4                    0.0                             NaN           0.0   

   new_deaths  new_deaths_smoothed  total_deaths_per_million  ...  \
0         NaN                  NaN                       NaN  ...   
1         NaN                  NaN                       NaN  ...   
2         NaN                  NaN                       NaN  ...   
3         0.0                  NaN                       0.0  ...   
4         0.0                  NaN                       0.0  ...   

   diabetes_prevalence  gdp_per_capita  stringency_index  \
0                 10.9     1516.273315               0.0   
1                 10.9     1516.273315               0.0   
2                 10.9     1516.273315               0.0   
3                 10.9     1516.273315               0.0   
4                 10.9     1516.273315               0.0   

   hospital_beds_per_thousand  handwashing_facilities  code  continent  \
0                        0.39               48.214695   AFG       Asia   
1                        0.39               48.214695   AFG       Asia   
2                        0.39               48.214695   AFG       Asia   
3                        0.39               48.214695   AFG       Asia   
4                        0.39               48.214695   AFG       Asia   

       country        date  population  
0  Afghanistan  2020-01-01  40578801.0  
1  Afghanistan  2020-01-02  40578801.0  
2  Afghanistan  2020-01-03  40578801.0  
3  Afghanistan  2020-01-04  40578801.0  
4  Afghanistan  2020-01-05  40578801.0  

[5 rows x 24 columns]

In [108]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 508524 entries, 0 to 508523
Data columns (total 24 columns):
 #   Column                           Non-Null Count   Dtype  
---  ------                           --------------   -----  
 0   total_cases                      493767 non-null  float64
 1   new_cases                        490849 non-null  float64
 2   total_cases_per_million          493767 non-null  float64
 3   new_cases_smoothed               489644 non-null  float64
 4   new_cases_per_million            490849 non-null  float64
 5   new_cases_smoothed_per_million   489644 non-null  float64
 6   total_deaths                     493767 non-null  float64
 7   new_deaths                       491630 non-null  float64
 8   new_deaths_smoothed              490417 non-null  float64
 9   total_deaths_per_million         493767 non-null  float64
 10  new_deaths_per_million           491630 non-null  float64
 11  new_deaths_smoothed_per_million  490417 non-null  float64
 12  po

In [109]:
df['date'] = pd.to_datetime(df['date'])

# Step 2: Group by 'date' and keep only complete rows
df_cleaned = (
    df.groupby('date')
      .apply(lambda group: group.dropna(how='any'))  # keep only rows with all values present
      .reset_index(drop=True)
)

/tmp/ipython-input-109-2122837832.py:6: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda group: group.dropna(how='any'))  # keep only rows with all values present


In [110]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70687 entries, 0 to 70686
Data columns (total 24 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   total_cases                      70687 non-null  float64       
 1   new_cases                        70687 non-null  float64       
 2   total_cases_per_million          70687 non-null  float64       
 3   new_cases_smoothed               70687 non-null  float64       
 4   new_cases_per_million            70687 non-null  float64       
 5   new_cases_smoothed_per_million   70687 non-null  float64       
 6   total_deaths                     70687 non-null  float64       
 7   new_deaths                       70687 non-null  float64       
 8   new_deaths_smoothed              70687 non-null  float64       
 9   total_deaths_per_million         70687 non-null  float64       
 10  new_deaths_per_million           70687 non-null  float64  

In [113]:
df_cleaned.head()

total_cases  new_cases  total_cases_per_million  new_cases_smoothed  \
0          0.0        0.0                      0.0                 0.0   
1          0.0        0.0                      0.0                 0.0   
2          0.0        0.0                      0.0                 0.0   
3          0.0        0.0                      0.0                 0.0   
4          0.0        0.0                      0.0                 0.0   

   new_cases_per_million  new_cases_smoothed_per_million  total_deaths  \
0                    0.0                             0.0           0.0   
1                    0.0                             0.0           0.0   
2                    0.0                             0.0           0.0   
3                    0.0                             0.0           0.0   
4                    0.0                             0.0           0.0   

   new_deaths  new_deaths_smoothed  total_deaths_per_million  ...  \
0         0.0                  0.0                       0.0  ...   
1         0.0                  0.0                       0.0  ...   
2         0.0                  0.0                       0.0  ...   
3         0.0                  0.0                       0.0  ...   
4         0.0                  0.0                       0.0  ...   

   diabetes_prevalence  gdp_per_capita  stringency_index  \
0                 10.9     1516.273315               0.0   
1                  7.1    11198.233398               0.0   
2                  5.6    15094.284180               0.0   
3                 11.3    51854.714844               0.0   
4                 14.2     6263.004395               0.0   

   hospital_beds_per_thousand  handwashing_facilities  code  continent  \
0                        0.39               48.214695   AFG       Asia   
1                        1.90               84.804100   DZA     Africa   
2                        4.82               89.304024   AZE       Asia   
3                        1.74              100.000000   BHR       Asia   
4                        0.79               61.728539   BGD       Asia   

       country       date   population  
0  Afghanistan 2020-01-09   40578801.0  
1      Algeria 2020-01-09   45477339.0  
2   Azerbaijan 2020-01-09   10295256.0  
3      Bahrain 2020-01-09    1533408.0  
4   Bangladesh 2020-01-09  169384838.0  

[5 rows x 24 columns]

In [115]:
# Number of unique countries
unique_countries = df_cleaned['country'].nunique()
print(f"Number of unique countries in cleaned data: {unique_countries}")

# Names of unique countries
country_names = df_cleaned['country'].unique()
print("Countries in cleaned data:")
for country in sorted(country_names):
    print("-", country)

Number of unique countries in cleaned data: 65
Countries in cleaned data:
- Afghanistan
- Algeria
- Azerbaijan
- Bahrain
- Bangladesh
- Barbados
- Belize
- Bhutan
- Bolivia
- Bosnia and Herzegovina
- Burundi
- Cambodia
- Central African Republic
- China
- Colombia
- Costa Rica
- Dominican Republic
- Ecuador
- Egypt
- El Salvador
- Eswatini
- Ethiopia
- Fiji
- Gambia
- Georgia
- Ghana
- Guatemala
- Guinea
- Guyana
- Haiti
- Honduras
- India
- Indonesia
- Iraq
- Jamaica
- Kazakhstan
- Kiribati
- Kyrgyzstan
- Laos
- Malawi
- Mexico
- Moldova
- Mongolia
- Mozambique
- Myanmar
- Nepal
- Niger
- Oman
- Pakistan
- Paraguay
- Philippines
- Serbia
- Solomon Islands
- Somalia
- Sri Lanka
- Sudan
- Suriname
- Tajikistan
- Togo
- Trinidad and Tobago
- Tunisia
- Turkmenistan
- Uzbekistan
- Vietnam
- Zimbabwe


In [117]:
# Get country sets
covid_countries = set(df_cleaned['country'].unique())
health_countries = set(health_df['Country'].unique())

# Find unmatched countries
unmatched = covid_countries - health_countries
print(f"\nUnmatched countries ({len(unmatched)}):")
for c in sorted(unmatched):
    print(c)



Unmatched countries (13):
BHUTAN
BOLIVIA
CENTRAL AFRICAN REPUBLIC
GUINEA
HONDURAS
LAOS
MOLDOVA
SOMALIA
SUDAN
SURINAME
TAJIKISTAN
TRINIDAD AND TOBAGO
VIETNAM


In [116]:
# Step 1: Load health data
health_df = pd.read_csv('/content/combined_health_data.csv')

# Step 2: Standardize capitalization in both datasets
df_cleaned['country'] = df_cleaned['country'].str.upper()
health_df['Country'] = health_df['Country'].str.upper()

# Step 3: Merge on 'country' and 'Country'
merged_df = pd.merge(
    df_cleaned,
    health_df,
    left_on='country',
    right_on='Country',
    how='inner'  # Only keep rows with matching country names
)

# Step 4: Drop redundant column if needed
merged_df = merged_df.drop(columns=['Country'])

# Step 5: Preview the result
print(f"Merged dataset shape: {merged_df.shape}")
print("Unique countries after merge:", merged_df['country'].nunique())
print(merged_df.head())


Merged dataset shape: (56546, 28)
Unique countries after merge: 52
   total_cases  new_cases  total_cases_per_million  new_cases_smoothed  \
0          0.0        0.0                      0.0                 0.0   
1          0.0        0.0                      0.0                 0.0   
2          0.0        0.0                      0.0                 0.0   
3          0.0        0.0                      0.0                 0.0   
4          0.0        0.0                      0.0                 0.0   

   new_cases_per_million  new_cases_smoothed_per_million  total_deaths  \
0                    0.0                             0.0           0.0   
1                    0.0                             0.0           0.0   
2                    0.0                             0.0           0.0   
3                    0.0                             0.0           0.0   
4                    0.0                             0.0           0.0   

   new_deaths  new_deaths_smoothed  total_d

In [118]:
# Optional: Save result
merged_df.to_csv('/content/final_covid_health_combined.csv', index=False)